### ISSUE TITLE
Memory extender applications reliably crash DOSBox

### ISSUE THREAD

**Issue by sarahwhitman** [dosbox-staging/dosbox-staging#746]
DOSBox staging now reliably crashes when attempting to run applications that depend on DOS4GW or DOS32A memory extenders (Warcraft 1 & 2, Death Rally, Dark Legions, Wacky Wheels, etc.) but operates as expected when running non-extender games (Cosmo's Cosmic Adventures, Jill of the Jungle, SimCity 1, etc.)

To reproduce, install fresh Windows x64 DOSBox-staging with plain, auto-generated config file and attempt to execute a program/game that uses a memory extender. I think memory extenders are the common thread here, but I could definitely be wrong!

The game should load but it does not - and if we dump the output to stdout, no errors are generated, the output halts mid sentence (mine usually cuts off with "DOSBox has switched to max")

I can reproduce this on both of my laptops:
1) Dell Precision 5520, i5 7440hq, Intel UHD 630 graphics (driver v.27.20.100.8935), Windows 10 1809 LTSC x64 (17763.1613)
2) HP EliteBook 1040 G1, i7 4600u, Intel HD 4400 graphics (driver v.20.19.15.5126), Windows 10 1607 LTSB x64 (14393.4046)

The crash occurs starting with build 'v0.76.0-alpha-1000-g9a77a9' and all subsequent builds. The crash does not occur on build '0.76.0.alpha.985.g6b84faa5' and all prior builds.

Happy to provide additional info if needed!

**Comment by kcgen** [comment]
Thanks @sarahwhitman !
Confirmed. It appears to be a Windows-specific issue with the dynamic core.
For an immediate work-around, launch `dosbox -editconf`, then set `core = normal`. 

![Uploading Capture.PNG…]()

**Comment by dreamer** [comment]
Uh, this is really weird - I cannot reproduce this issue with debug builds or even with release builds compiled locally…

But it does indeed happen with builds generated by our CI. :/

**Comment by kcgen** [comment]
Swap out the package with your build? Or take down the windows package and mention its been pulled due to unexpected build regression?
I think it's severe enough that we don't want to waste people's time with it.

**Comment by dreamer** [comment]
But if this happens then I have no idea if my local build is going to work on other people's computers... I am trying to pinpoint commit that introduced regression (we did not have any risky CPU changes in this time period).

**Comment by kcgen** [comment]
What we do know is the posted package has failed on 4 out of 4 separate Windows 10 PCs with 100% reliability; so I think we need to pull it regardless of what the corrective strategy will be.

I also fully agree: find where is went off the rails is key to solving it; but it's separate from mitigating the current damage. Unless you think it's going to be quick to find and can swap in a working package. 

My hunch is Windows CI might be busted for another week or so if the current VS2019 has some subtle bug that's giving us some flawed machine code (which doesn't exit in your revision).

**Comment by dreamer** [comment]
OK, there is some progress now...

Previously I was using Visual Studio 16.7.4 - in that one, I was unable to reproduce crash at all (neither in Release or Debug profile). But I use Windows rarely, so I haven't noticed there was an "upgrade". I just upgraded to 16.8.2 - and now our Relase build *do* crash (only Release builds, Debug builds are fine - grrr).

@sarahwhitman Thank you so much about mentioning that this changed after `0.76.0-alpha-985-g6b84faa5` - I was able to check it a bit further:
- `dosbox-staging-windows-x64-v0.76.0-alpha-992-gfa481` - this build crashes
- `dosbox-staging-windows-x64-v0.76.0-alpha-990-g780b` - this build still works

In `v0.76.0-alpha-992-gfa481` (fa481b3db413d7131cdb852eb73adc981d85f3fd) we updated VC_REDIST libraries only (no change in code) - but this is *not* the cause of the problem (I tried using older dlls and our new exe still crash). I *think* this VC_REDIST upgrade coincided with the Visual Studio / MSVC version bump.

So, I rebuilt v0.75.2 on my local machine with new Visual Studio... and it crashes as well.

I did read through [release notes for this Visual Studio 16.8.x](https://docs.microsoft.com/en-gb/visualstudio/releases/2019/release-notes#--visual-studio-2019-version-1682-) and there's nothing of interest listed there :/ (perhaps I should look at changes in MSVC or something).

**TL;DR;** I am 95% sure Microsoft changed / triggered something in new VS and now it suddenly triggers our W^X problems (#253) - but **only when building for x64**.

@sarahwhitman @kcgen can you download [32-bit Windows build](https://github.com/dosbox-staging/dosbox-staging/releases/download/v0.76.0/dosbox-staging-windows-x86-v0.76.0.zip) and confirm it does **not** crash?

**Comment by dreamer** [comment]
@Diduz confirmed that 32-bit builds still work correctly in https://github.com/dosbox-staging/dosbox-staging/issues/742#issuecomment-738692324 - so we have 2 machines where 32-bit build works, and 4 where 64-bit does not.

I am taking down our 64-bit builds and adding a note to "Known issues" in the release notes.

**Comment by dreamer** [comment]
I just realized that it's really weird that W^X would be happening only for 64-bit builds and not for 32-bit ones… This might be a bug in 64-bit dynamic recompiler, that is triggered in MSVC builds only… So this *might not* be a duplicate of #253 - we don't know yet.

**Comment by sarahwhitman** [comment]
@dreamer Wow, you were able to track that down quickly! Confirming that the Windows 32-bit 0.76.0 release does not crash on both of my test machines, and neither does v0.77.0-alpha-3-g973e

**Comment by kcgen** [comment]
Nice work in narrowing this down and all the help everyone's providing! 🚀  (Back to 😴 for me now)

**Comment by MasterO2** [comment]
@kcgen @dreamer Any progress on this issue?

**Comment by dreamer** [comment]
We know the root cause, someone needs to port the fix from DOSBox-X and test it. I can't do it, because I don't have access to Windows machine for next few weeks.

**Comment by sarahwhitman** [comment]
The linked PR resolves this issue on my end! All of the mentioned games (Warcraft 1 & 2, Death Rally, Dark Legions, Wacky Wheels) no longer crash on my system, even when manually setting core=dynamic - thank you @kcgen !
